modules/userId/index.js

```diff
@@ -453,7 +453,7 @@ function getSubmoduleId(submodules, sourceName) {
     return {};
   }
   const submodule = submodules.filter(sub => isPlainObject(sub.idObj) &&
-    Object.keys(sub.idObj).length && USER_IDS_CONFIG[Object.keys(sub.idObj)[0]]?.source === sourceName);
+    Object.keys(sub.idObj).length && USER_IDS_CONFIG[Object.keys(sub.idObj)[0]]?.find(conf => conf.source === sourceName));
```

Comment:
actually, since this is only used in here: 

https://github.com/prebid/Prebid.js/blob/d1b228c4a6e4cf6f5fa8a7899c51acbb5a5f4115/modules/userId/index.js#L656-L658

I think what we should do instead is reverse the filtering - instead of finding the module for a given source to get its EIDs, it makes more sense now to get all EIDs and filter them by source (since they can come from different modules).

Reply:
exactly, "get all EIDs and filter them by source". If there is a collision, i.e., 2 or more eids with the same source check the provider and choose the one where the provider and source are the same. The question then becomes what do we do when one eid has no provider value?

Reply:
The `eids#USER_IDS_CONFIG` could be adjusted to always have a global `provider` for each user id that is then used as `provider` for all `uids` extracted based on it. That way every `eid/uid` would have a `provider`.  For example:

```
  'lipb': {
  	provider: 'liveintent.com',
  	extractors: [
	    // liveIntentId
	    {
	      getValue: function(data) {
	        return data.lipbid;
	      },
	      source: 'liveintent.com',
	      atype: 3,
	      getEidExt: function(data) {
	        if (Array.isArray(data.segments) && data.segments.length) {
	          return {
	            segments: data.segments
	          };
	        }
	      }
	    },

	    // uid2
	    {
	      source: 'uidapi.com',
	      atype: 3,
	      getValue: function(data) {
	        return data.uid2;
	      }
	    }
	  ]
	}
```
And `eids#createEidObjects` would take care of building the `uid.ext` using the global `provider` value and output 

```
[{
    source: 'liveintent.com',
    uids: [{
        id: 'some-random-id-value',
        atype: 1
    }],
    ext: {
        provider: 'liveintent.com',
        segments: ['s1', 's2']
    }
},
{
    source: 'uidapi.com',
    uids: [{
        id: 'some-random-id-value',
        atype: 3,
        ext: { 
            provider: 'liveintent.com' 
        }
    }]
}]
```

Reply:
One related question: what if there is a collision and neither of the providers is the owner of the id? 

Maybe it would make sense to return an `eid` with multiple `uids` when `getUserIdsAsEidBySource` is called and there is a collision?

For example: `getUserIdsAsEidBySource('uidapi.com')` would then return

```
[{
    source: 'uidapi.com',
    uids: [{
        id: 'some-random-id-value',
        atype: 3,
        ext: { provider: 'uidapi.com' }
    },
    {
        id: 'some-random-id-value',
        atype: 3,
        ext: { provider: 'liveintent.com' }
    }]
}]
```
when a UID2 could be resolved via LI's and the UID2 adapter. The caller then decides which `uid` to pick if there are multiple. We also can make sure that the `uid` of the id owner is always at index `0` in the `uids` array.

Reply:
> currently getUserIdsAsEidBySource returns a single object, and it's part of the public API, so that's not really up for change. if there's more than one EID we have to pick one. 

I agree that the API should not be changed in this regard. `getUserIdsAsEidBySource` should return one single object - one `eid`. An `eid`, however, can have an array of `uids`. My proposal is rather about - in case of collisions - exposing multiple `uids` with their providers within the `eid` that is returned. By providing the "provider" field the `uids` become distinguishable and the caller can decide what id to use in case there are multiple based on the provider.

Reply:
> Can we make this configurable? Can the pub declare a list of priorities in the core user id module?
@jdwieland8282 Sounds good. 

@jdwieland8282 What do you think, should that config allow to provide priority per identifier type - e.g. `uid2` from `uidapi.com` has higher priority than `uid2` from `liveintent.com`? Or should it be a config that defines that all identifiers from one user id module have priority over same types of identifiers exposed by any other module with lower priority? I personally tend towards the first option as it is more fine grained. 

Also, I would apply this prioritization logic when building the `eids` array, not only in `getUserIdsAsEidBySource`. Are we on the same page here?